src/ReinforcementLearningDatasets/src/d4rl/d4rl_dataset.jl

```diff
@@ -0,0 +1,141 @@
+const SARTS = (:state, :action, :reward, :terminals, :next_state)
+const SART = (:state, :action, :reward, :terminals)
+
+# not exporting D4RLDataSet. (no need outside) -> Maybe make a constructor for D4RLDataSet that will return a D4RLDataSet dataset
+# write docstring for this.
+struct D4RLDataSet{T<:AbstractRNG}
+    dataset::Dict{Symbol, Any}
+    size::Int
+    rng::T
+    batch_size::Int64
+    style::Tuple
+    meta::Dict
+end
+
+# haven't included all the functionality like is_sequential, will be implemented soon
+# Maybe enable the users providing their own paths to datasets if they already have it
+# add an optional env_arg for additional checking
+"""
+    dataset(dataset::String; style::Tuple, rng<:AbstractRNG, is_shuffle::Bool, max_iters::Int64, batch_size::Int64)
+
+Creates a dataset of enclosed in a D4RLDataSet type and other related metadata for the `dataset` that is passed.
+The dataset type is an iterable that fetches batches when used in a for loop for convenience during offline training.
+
+`dataset`: Name of the D4RLDataSet dataset.
+`style`: the style of the iterator and the Dict inside D4RLDataSet that is returned.
+`rng`: StableRNG
+`max_iters`: maximum number of iterations for the iterator.
+`is_shuffle`: whether the dataset is shuffled or not. `true` by default.
+`batch_size`: batch_size that is yielded by the iterator. Defaults to 256.
+
+The returned type is an infinite iterator which can be called using `iterate` and will return batches as specified in the dataset.
+"""
+function dataset(dataset::String;
+    style=SARTS, 
+    rng = StableRNG(123), 
+    is_shuffle = true, 
+    batch_size=256
+)
+    
+    try 
+        @datadep_str "d4rl-"*dataset 
+    catch 
+        throw("The provided dataset is not available") 
+    end
+        
+    path = @datadep_str "d4rl-"*dataset
+
+    @assert length(readdir(path)) == 1
+    file_name = readdir(path)[1]
+    
+    # try to incorporate multi threading in loading stuff
+    data = h5open(path*"/"*file_name, "r") do file
+        read(file)
+    end
+
+    #check if we could obtain more info so that we could verify if the dataset actually matches the requirement
+    #do some sanity checks on the data
+    verify(data)
+
+    dataset = Dict{Symbol, Any}()
+    meta = Dict{String, Any}()
+
+    N_samples = size(data["terminals"])[1]
+    
+    for (key, d_key) in zip(["observations", "actions", "rewards", "terminals"], Symbol.(["state", "action", "reward", "terminal"]))
+            dataset[d_key] = data[key]
+    end
+    
+    if style == SARTS
+        dataset[:next_state] = @view data["observations"][:, 2:N_samples]
+        dataset[:next_state] = cat(dataset[:next_state], data["observations"][:, 1]; dims = 2)
+    end
+    
+    # put the data in separate container based on requirements
+    for key in keys(data)
+        if !(key in ["observations", "actions", "rewards", "terminals"])
+            meta[key] = data[key]
+        end
+    end
+    
+    if is_shuffle
```

Comment:
This is not necessary in constructor now?

Reply:
Are you meaning meta[key]? It is just for obtaining more information that is available about the dataset. And it doesn't hurt to have.

Any specific reason as to why it need not be present?

Reply:
Sorry. I mean these lines:

https://github.com/JuliaReinforcementLearning/ReinforcementLearning.jl/pull/403/files/5d1687b6f109b24f531172ea96a4dd1b9cdf8ac9#diff-e4c3688b8152572a615135dec3f6bc48e095f115bcc68fedd22fabd0f9a464e1R91-R108

Reply:
I think we agreed that the `is_shuffle` stuff only occurs in `iterate`?

Reply:
Fair enough.

Reply:
So if I modify iterate based on shuffle bool it would do?

Reply:
Sorry for missing that detail.

Reply:
> So if I modify iterate based on shuffle bool it would do?

is_shuffle is used to control whether we use the `rng` to select baches randomly.

Reply:
Have made some fixes